### ISSUE TITLE
[API Proposal]: Add support for Ed25519

### ISSUE THREAD

**Issue by annerajb** [dotnet/runtime#63174]
## Background

I work a lot with Ed25519 and love dotnet to do development, Last year I reported a issue trying to use Ed25519 and the TLS 1.3 integration wit openssl on SslStream did not support Ed25519 certificates and keys. [See here](https://github.com/dotnet/runtime/issues/46513) 

I hope this proposal is more complete and actionable to get any process started.

## Traction / usage

This has been requested for a while now or will be used more in the future.
* This was set forth as a proposal all the way back in [2015 here ](https://github.com/dotnet/runtime/issues/14741)
* This will now be included on the upcoming FIPS 186-5 standard allowing it's usage on FIPS enabled systems.
* It's also used in SSH keys and becoming more and more common making it beneficial for the dotnet ecosystem as a whole to support this suite of algorithms.
* [DNSSec](https://ed25519.nl/) with RFC 8080 adds ed448 and ed25519 as part of the signing.

## Reference / resources:

[DNS Sec ed25519 ](https://ed25519.nl/)
[RFC 8080: Edwards-Curve Digital Security Algorithm (EdDSA) for DNSSEC](https://datatracker.ietf.org/doc/html/rfc8080)
[RFC 8032: Edwards-Curve Digital Signature Algorithm (EdDSA)](https://datatracker.ietf.org/doc/html/rfc8032)
[RFC 8410: Algorithm Identifiers for Ed25519, Ed448, X25519, and X448
        for Use in the Internet X.509 Public Key Infrastructure](https://datatracker.ietf.org/doc/html/rfc8410#section-8)

### API Proposal

Ed25519 is a specific set of options from Edwards DSA (EdDSA).

Due to lack of general support for EdDSA from underlying stacks, this proposal ignores the general algorithm, and focuses on this specific variant.

```C#

namespace System.Security.Cryptography
{
    public abstract partial class Ed25519 : System.IDisposable
    {
        //  RFC 8032
        public const int SignatureSizeInBytes = 64;
        public const int PrivateKeySizeInBytes = 32;
        public const int PublicKeySizeInBytes = 32;

        protected Ed25519();

        public static bool IsSupported { get; }

        public static Ed25519 GenerateKey();

        public static Ed25519 ImportPublicKey(byte[] source);
        public static Ed25519 ImportPublicKey(ReadOnlySpan<byte> source);

        public static Ed25519 ImportPrivateKey(byte[] source);
        public static Ed25519 ImportPrivateKey(ReadOnlySpan<byte> source);

        public byte[] ExportPublicKey();
        public void ExportPublicKey(Span<byte> destination);
        protected abstract void ExportPublicKeyCore(Span<byte> destination);

        public byte[] ExportPrivateKey();
        public void ExportPrivateKey(Span<byte> destination);
        protected abstract void ExportPrivateKeyCore(Span<byte> destination);

        public byte[] SignData(byte[] data);
        public void SignData(ReadOnlySpan<byte> data, Span<byte> destination);
        protected abstract void SignDataCore(ReadOnlySpan<byte> data, Span<byte> destination);

        public bool VerifyData(byte[] data, byte[] signature);
        public bool VerifyData(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature);
        protected abstract bool VerifyDataCore(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature);

        public void Dispose();
        protected virtual void Dispose(bool disposing);

        public bool TryExportEncryptedPkcs8PrivateKey(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters, Span<byte> destination, out int bytesWritten);
        public bool TryExportEncryptedPkcs8PrivateKey(ReadOnlySpan<char> password, PbeParameters pbeParameters, Span<byte> destination, out int bytesWritten);
        public bool TryExportEncryptedPkcs8PrivateKey(string password, PbeParameters pbeParameters, Span<byte> destination, out int bytesWritten);

        public byte[] ExportEncryptedPkcs8PrivateKey(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters);
        public byte[] ExportEncryptedPkcs8PrivateKey(ReadOnlySpan<char> password, PbeParameters pbeParameters);
        public byte[] ExportEncryptedPkcs8PrivateKey(string password, PbeParameters pbeParameters);

        public string ExportEncryptedPkcs8PrivateKeyPem(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters);
        public string ExportEncryptedPkcs8PrivateKeyPem(ReadOnlySpan<char> password, PbeParameters pbeParameters);
        public string ExportEncryptedPkcs8PrivateKeyPem(string password, PbeParameters pbeParameters);

        public byte[] ExportPkcs8PrivateKey();
        public string ExportPkcs8PrivateKeyPem();
        public bool TryExportPkcs8PrivateKey(Span<byte> destination, out int bytesWritten);
        protected abstract bool TryExportPkcs8PrivateKeyCore(Span<byte> destination, out int bytesWritten);

        public byte[] ExportSubjectPublicKeyInfo();
        public bool TryExportSubjectPublicKeyInfo(Span<byte> destination, out int bytesWritten);
        public string ExportSubjectPublicKeyInfoPem();

        public static Ed25519 ImportEncryptedPkcs8PrivateKey(ReadOnlySpan<byte> passwordBytes, ReadOnlySpan<byte> source);
        public static Ed25519 ImportEncryptedPkcs8PrivateKey(ReadOnlySpan<char> password, ReadOnlySpan<byte> source);
        public static Ed25519 ImportEncryptedPkcs8PrivateKey(string password, byte[] source);

        public static Ed25519 ImportFromEncryptedPem(ReadOnlySpan<char> source, ReadOnlySpan<byte> passwordBytes);
        public static Ed25519 ImportFromEncryptedPem(string source, byte[] passwordBytes);
        public static Ed25519 ImportFromEncryptedPem(ReadOnlySpan<char> source, ReadOnlySpan<char> password);
        public static Ed25519 ImportFromEncryptedPem(string source, string password);

        public static Ed25519 ImportFromPem(ReadOnlySpan<char> source);
        public static Ed25519 ImportFromPem(string source);

        public static Ed25519 ImportPkcs8PrivateKey(byte[] source);
        public static Ed25519 ImportPkcs8PrivateKey(ReadOnlySpan<byte> source);

        public static Ed25519 ImportSubjectPublicKeyInfo(byte[] source);
        public static Ed25519 ImportSubjectPublicKeyInfo(ReadOnlySpan<byte> source);
    }

    public sealed partial class Ed25519OpenSsl : Ed25519
    {
        [UnsupportedOSPlatformAttribute("android")]
        [UnsupportedOSPlatformAttribute("browser")]
        [UnsupportedOSPlatformAttribute("ios")]
        [UnsupportedOSPlatformAttribute("osx")]
        [UnsupportedOSPlatformAttribute("tvos")]
        [UnsupportedOSPlatformAttribute("windows")]
        public MLDsaOpenSsl(SafeEvpPKeyHandle pkeyHandle);

        public SafeEvpPKeyHandle DuplicateKeyHandle();
    }
}
```

**Comment by hamarb123** [comment]
I'm certainly not an expert on cryptography, but I do know about macOS interop:
You could always write an obj-c wrapper around CryptoKit and then use that in .NET through P/Invoke (wrapper could even be exported as c functions instead of obj-c functions for easier P/Invoke), this would probably not be too much effort for 1 set of APIs - https://stackoverflow.com/questions/65686717/can-cryptokit-be-used-with-objective-c, https://github.com/xamarin/xamarin-macios/issues/6239. Then again, it may be better to wait for `xamarin-macios`'s swift binding library solution, but I'm sure it would essentially be an easy drop-in if we were to change.

**Comment by vcsjones** [comment]
> You could always write an obj-c wrapper around CryptoKit and then use that in .NET through P/Invoke (wrapper could even be exported as c functions instead of obj-c functions for easier P/Invoke)

We've had many discussions about CryptoKit (https://github.com/dotnet/runtime/issues/52482#issuecomment-838290365, https://github.com/dotnet/runtime/issues/29811) and it's not entirely straight forward.

Even if CryptoKit were used, that would only give capabilities for the raw Ed25519 primitive. It would not work with `X509Certificate{2}` because, the best of my knowledge, [Key Services](https://developer.apple.com/documentation/security/certificate_key_and_trust_services) do not support Ed25519 and nor does macOS support [RFC8410](https://datatracker.ietf.org/doc/html/rfc8410), in that macOS will reject X.509 certificates that contain an id-Ed25519 AlgorithmIdentifier.


The lack of platform support anywhere other than OpenSSL currently makes this a bit of a tough sell to me. I have some thoughts on the APIs

1. Generally I think .NET is starting to shift away from the `SymmetricAlgorithm` and `AsymmetricAlgorithm` classes (e.g. `ChaCha20Poly1305` in .NET 6). I don't think this class should inherit from `AsymmetricAlgorithm`. These base classes are an artifact of an older "crypto agility" design where everything had a common base class.

    We could clean up the API shape then. For example, instead of `byte[] SignData(byte[] data, int offset, int count);` We could just have `SignData(ReadOnlySpan<byte> data)`. Same for `VerifyData`, etc. The `offset` / `count` is pre-span era.

2. How would someone generate a Ed448 (if it exists in the future) with this API shape? I wonder if `Create` should accept `ECCurveType` so that, if in the future, more EdDSA algorithms are added, only the enum will need to be changed.

3.

**Comment by annerajb** [comment]
> 1. Generally I think .NET is starting to shift away from the `SymmetricAlgorithm` and `AsymmetricAlgorithm` classes (e.g. `ChaCha20Poly1305` in .NET 6). I don't think this class should inherit from `AsymmetricAlgorithm`. These base classes are an artifact of an older "crypto agility" design where everything had a common base class.

One of the initial benefit I saw from using `AsymmetricAlgorithm` as a base class was all the function it already contained to handle Pkcs8, SubjectPublicKey, Pem Loading some of this [which where recently added to it](https://github.com/dotnet/runtime/commit/20997528d84805565217ba8a85133af507144065#diff-9672ec44dca6df47af54b9cd23e9fe8161ccc5f0e9c03a14bf82be1eeb32458e). 

On a closer look it seems the heavy lifting is really handled by `PemKeyHelpers`.

>    We could clean up the API shape then. For example, instead of `byte[] SignData(byte[] data, int offset, int count);` We could just have `SignData(ReadOnlySpan<byte> data)`. Same for `VerifyData`, etc. The `offset` / `count` is pre-span era.

Updated the sample to switch to ReadOnlySpan and remove offset arguments.

> 2. How would someone generate a Ed448 (if it exists in the future) with this API shape? I wonder if `Create` should accept `ECCurveType` so that, if in the future, more EdDSA algorithms are added, only the enum will need to be changed.

I noticed I missed renaming `ECCurveType` to be `EDCurveType` went ahead and did that change.

I also removed the parameter less `Create` static function and instead added a new one that accepts the curve type.

`public static new EDDsa Create(EDCurveType curve);`

Also added a `public static bool IsSupported { get; }` since it would be needed anyways

**Comment by bartonjs** [comment]
> Generally I think .NET is starting to shift away from the SymmetricAlgorithm and AsymmetricAlgorithm classes

For SymmetricAlgorithm that's effectively true... not because of a strong desire to abandon the types, but because all we've added lately was AE(AD) and it has such a different shape than SymmetricAlgorithm and ICryptoTransform that we decided they were just their own thing.  (We didn't make a new AE(AD) base type because the algorithms have enough variation that the base type wasn't providing value).

For AsymmetricAlgorithm, we've been increasing the places that can take one... like EnvelopedCms.Decrypt.  It's just trickier in API because DSA and ECDSA don't need extra context, but RSA has the PKCS1 vs PSS or PKCS1 vs OAEP (and what OAEP) split... which is why CertificateRequest doesn't take AsymmetricAlgorithm.

> I also removed the parameter less Create static function and instead added a new one that accepts the curve type.

Why remove the old one?  If you're loading from an existing key then having to specify a parameter just makes things potentially misleading (e.g. it says "create Ed25519", but after the load it's Ed448).  An overload that takes the parameter is fine, that tells the implicit key generation which curve to use.

> `public enum EDCurveType`

Using an enum is very limiting.  If we load an Ed448 from a key file but that's not in the enum we're in a bad state.  It probably wants to be something like a strongly typed string.

> What should the casing / naming be? EdDsa or EDDsa?

Either "EdwardsDsa" or "EdDsa", but definitely not "EDDsa"... the first D doesn't stand for anything, it's just "Ed" or "Edwards" concat (Digital Signature Algorithm)=>DSA=>Dsa.

> This would only support platforms where Openssl 1.1.1 is used by dotnet

https://dotnet.microsoft.com/en-us/platform/telemetry says that 60% of the command line usage is now on Linux (which is higher than I expected... guess we did a good job on cross-plat!).  Assuming that all Linux usages have OpenSSL 1.1.1 or OpenSSL 3, and that none of them have disabled EdDSA, that leaves a 60/40 split that we don't have any current expectation of healing. 

If macOS supported certificates with EdDSA then we could at least assume/pretend that Windows will _eventually_ follow; and if Windows were to support it in preview builds then we'd know that eventually 38% (the 98% that install updates of the 39% Windows usage) more would have `IsSupported=>true`, and we'd be at 98/2 (and then Apple might come on board and we get 99/1).

Right now having something that only works on one OS family is a bit of a hard sell.

**Comment by vcsjones** [comment]
> For AsymmetricAlgorithm, we've been increasing the places that can take one... like EnvelopedCms.Decrypt. It's just trickier in API because DSA and ECDSA don't need extra context, but RSA has the PKCS1 vs PSS or PKCS1 vs OAEP (and what OAEP) split... which is why CertificateRequest doesn't take AsymmetricAlgorithm.

I think that's kind of what bugs me a little about it. It's a cryptoagility-esq API that doesn't quite work. For that particular example, even though it accepts an `AsymmetricAlgorithm` it's going to throw for anything other than RSA (and Windows gives special treatment for FFDH).

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.Decrypt.cs#L47-L53

EdDSA can't encrypt, but you'll be able to pass an instance in anyway. Maybe X25519 could if we ever added something like ECIES if we ever got there.

And because `AsymmetricAlgorithm` doesn't actually have sign, encrypt, or anything on it, you can't rely on the polymorphism to "just work" for actual asymmetric operations.

**Comment by bartonjs** [comment]
Well, in your quoted example, we already know that we're a KeyTrans, which means that the relevant cert/KeyRecipientInfo was RSA (the only algorithm that would have matched a KeyTrans).  For KeyAgree, if we supported that, and we supported both DH and ECDH, it'd look like

```C#
if (privateKey is DiffieHellman dh)
{
}
else if (privateKey is ECDiffieHellman ecdh)
{
}
else
{
    throw something;
}
```

We could make the public entrypoint be `Decrypt(RecipientInfo, RSA)`, `Decrypt(RecipientInfo, DiffieHellman)`, `Decrypt(RecipientInfo, ECDiffieHellman)`... or just `Decrypt(RecipientInfo, AsymmetricAlgorithm)` and you have to give us the right key (since they're all going to fail for "the recipientinfo and key don't agree" and/or "I tried this key, but it didn't work")

Encrypt is harder, since there's not already a RecipientInfo (etc) to tell us if we're supposed to use PKCS1/OAEP-SHA1/OAEP-SHA-2-256/...

I agree it's a bit asymmetric (semi-pun not intended, but it's the best word), but it feels better than `object` or unnecessary overload explosion.

**Comment by mregen** [comment]
Imho this discussion should also include the support for `X509Chain` validation for EdDSA/Curve25519/Curve448 certs. 
I'd love to see the support for the EdDsa based certificates, but in our use case with the OPC UA protocol without PKI chaining and signature validation by the platform we can at best just use self signed.

> Right now having something that only works on one OS family is a bit of a hard sell.

There are many use cases in the industrial space where this makes a lot of sense. The edge devices run Linux and embedded sensors may only support the EdDsa/Curve25519/ChaChaPoly flavors due to their minimal CPUs.
With the recent push around [cybersecurity](https://www.federalregister.gov/documents/2021/05/17/2021-10460/improving-the-nations-cybersecurity) .NET may put itself on the sidelines if it doesn't support it...

**Comment by Fabi** [comment]
Still not implemented???

**Comment by Blackclaws** [comment]
Seeing how JWTs with EdDSA are becoming more popular this is definitely something that is needed sooner rather than later if we want to keep verifying those without having to rely on third party libraries.

**Comment by vcsjones** [comment]
@bartonjs 

As a first step, I did a rough implementation of Ed25519 for macOS and OpenSSL, enough to get passing with DJB's test vectors.

This meets our "two of three" semi-requirement, but it would mean we lack the ability to implement this on Windows.

Here is the API shape that I came up with:

```C#
namespace System.Security.Cryptography
{
    public abstract partial class Ed25519 : System.Security.Cryptography.AsymmetricAlgorithm
    {
        public int SignatureSizeInBytes { get; }
        public abstract bool HasPrivateKey { get; }

        [UnsupportedOSPlatformAttribute("android")] // MAYBE this does not apply.
        [UnsupportedOSPlatformAttribute("browser")]
        [UnsupportedOSPlatformAttribute("ios")]
        [UnsupportedOSPlatformAttribute("maccatalyst")]
        [UnsupportedOSPlatformAttribute("tvos")]
        [UnsupportedOSPlatformAttribute("windows")]
        public static new System.Security.Cryptography.Ed25519 Create();
        
        public byte[] ExportPrivateKey();
        public int ExportPrivateKey(System.Span<byte> destination);
        public bool TryExportPrivateKey(System.Span<byte> destination, out int bytesWritten);
        protected abstract int ExportPrivateKeyCore(System.Span<byte> destination);

        public byte[] ExportPublicKey();
        public int ExportPublicKey(System.Span<byte> destination);
        public bool TryExportPublicKey(System.Span<byte> destination, out int bytesWritten);
        protected abstract int ExportPublicKeyCore(System.Span<byte> destination);

        public abstract void GenerateKey();

        public void ImportPrivateKey(byte[] privateKey) { }
        public void ImportPrivateKey(System.ReadOnlySpan<byte> privateKey) { }
        protected abstract void ImportPrivateKeyCore(System.ReadOnlySpan<byte> privateKey);

        public void ImportPublicKey(byte[] publicKey) { }
        public void ImportPublicKey(System.ReadOnlySpan<byte> publicKey) { }
        protected abstract void ImportPublicKeyCore(System.ReadOnlySpan<byte> publicKey);

        public byte[] SignData(byte[] data);
        public byte[] SignData(System.ReadOnlySpan<byte> data);
        public int SignData(System.ReadOnlySpan<byte> data, System.Span<byte> destination);
        public bool TrySignData(System.ReadOnlySpan<byte> data, System.Span<byte> destination, out int bytesWritten);
        protected abstract int SignDataCore(System.ReadOnlySpan<byte> data, System.Span<byte> destination);

        public bool VerifyData(byte[] data, byte[] signature);
        public bool VerifyData(System.ReadOnlySpan<byte> data, System.ReadOnlySpan<byte> signature);
        protected abstract bool VerifyDataCore(System.ReadOnlySpan<byte> data, System.ReadOnlySpan<byte> signature);

        // Not included: All of the overrides from AsymmetricAlgorithm
    }

    public sealed partial class Ed25519OpenSsl : System.Security.Cryptography.Ed25519
    {
        [UnsupportedOSPlatformAttribute("android")]
        [UnsupportedOSPlatformAttribute("browser")]
        [UnsupportedOSPlatformAttribute("ios")]
        [UnsupportedOSPlatformAttribute("tvos")]
        [UnsupportedOSPlatformAttribute("windows")]
        public Ed25519OpenSsl(System.Security.Cryptography.SafeEvpPKeyHandle pkeyHandle);

        // Not included: All of the overrides from AsymmetricAlgorithm or Ed25519
    }
}
```

Some comparisons with the proposal above:

1. To keep mine simpler to read, all of the overrides from `AsymmetricAlgorithm` I removed. However yes, consider that we will override things like PKCS8 and SPKI exports.

2. No `EdDsa` type. The algorithm is the type. If we ever get to adding something like Ed448, we will add it as another type. If it proves to be useful to have an `EdDsa` "intermediate" abstract class, we can extract it out later like we did with `ECAlgorithm`.

3. The type directly exposes key imports and exports for the opaque keys. Since keys are opaque and by design don't have structure (like a Q.X and Q.Y) I saw no obvious need for a public / private tuple. Similarly, Ed25519 more strongly implies that having the private key without the public key is enough. So most key transport won't carry both (like ECDSA likes to carry D and Q together), the keys are just "sequences of bytes".

4. I'm more uncertain about the work in `X509Certificate2` and `X509Chain`. That's where our "two of three" becomes a "one of three". Apple implements the primitive only in CryptoKit. We can always do this as follow up work, but the situation for just the primitives is more clear as something that seems feasible to move forward with.

5. We _might_ be able to do Android. If I am reading Android's docs correctly, they implemented it in API Level 33, which is very new. The most recent, at the time of writing. We would need to do some kind of light-up work for it.

6. I went back and forth on "should there be an `Ed25519OpenSsl` class that is public. I went with "yes" to match the existing landscape. Note that the `IntPtr` constructor is not needed because Ed25519 works exclusively off of EVP_PKEY.

**Comment by bartonjs** [comment]
@vcsjones The original proposal reminded me that we should do the API for X509Certificate2+Ed25519, so I've gone ahead and merged that with your proposal into the top post.

Even if we cut it apart into two implementation PRs, it's probably worth getting the X.509 part pre-signed-off.

**Comment by vcsjones** [comment]
> The original proposal reminded me that we should do the API for X509Certificate2+Ed25519, so I've gone ahead and merged that with your proposal into the top post.

That's fine if we just want to get the API approved... but I don't know that we should do that for .NET 8. From my earlier post:

> I'm more uncertain about the work in X509Certificate2 and X509Chain. That's where our "two of three" becomes a "one of three". Apple implements the primitive only in CryptoKit. We can always do this as follow up work, but the situation for just the primitives is more clear as something that seems feasible to move forward with.

We can't implement Ed25519 + X.509 on Apple (and Apple summarily explodes on an Ed25519 cert).

Would you take that for .NET 8 if we can only implement those APIs for OpenSSL?

**Comment by bartonjs** [comment]
> Would you take that for .NET 8 if we can only implement those APIs for OpenSSL?

Assuming that an Ed25519 (public only) cert already opens on Linux, it feels reasonable to add GetPublicKey with the algorithm... and then to add Ed25519 to the PFX association table and add GetPrivateKey and CopyWithPrivateKey.  I think it'd be in good* company with DSA FIPS 186-3 in that I believe that "big DSA" certs work on Linux-only.

If loading the cert fails, then all bets are off, of course.

<sub>* debatable.  OK, not really debatable, it's bad company, but it's a stock English phrase...</sub>

**Comment by terrajobst** [comment]
* The `IsSupported` properties should use `[UnsupportedOSPlatformGuard]` annotations

```C#
namespace System.Security.Cryptography;

public abstract partial class Ed25519 : AsymmetricAlgorithm
{
    public const int SignatureSizeInBytes = 64;

    [UnsupportedOSPlatformGuard("android")]
    [UnsupportedOSPlatformGuard("browser")]
    [UnsupportedOSPlatformGuard("ios")]
    [UnsupportedOSPlatformGuard("tvos")]
    [UnsupportedOSPlatformGuard("windows")]
    public static bool IsSupported { get; }

    [UnsupportedOSPlatform("android")]
    [UnsupportedOSPlatform("browser")]
    [UnsupportedOSPlatform("ios")]
    [UnsupportedOSPlatform("tvos")]
    [UnsupportedOSPlatform("windows")]
    public static new Ed25519 Create();

    protected Ed25519();
    public abstract bool HasPrivateKey { get; }

    public byte[] ExportPrivateKey();
    public int ExportPrivateKey(Span<byte> destination);
    public bool TryExportPrivateKey(Span<byte> destination, out int bytesWritten);
    protected abstract int ExportPrivateKeyCore(Span<byte> destination);
    public byte[] ExportPublicKey();
    public int ExportPublicKey(Span<byte> destination);
    public bool TryExportPublicKey(Span<byte> destination, out int bytesWritten);
    protected abstract int ExportPublicKeyCore(Span<byte> destination);

    public abstract void GenerateKey();

    public void ImportPrivateKey(byte[] privateKey);
    public void ImportPrivateKey(ReadOnlySpan<byte> privateKey);
    protected abstract void ImportPrivateKeyCore(ReadOnlySpan<byte> privateKey);

    public void ImportPublicKey(byte[] publicKey);
    public void ImportPublicKey(ReadOnlySpan<byte> publicKey);
    protected abstract void ImportPublicKeyCore(ReadOnlySpan<byte> publicKey);
    public byte[] SignData(byte[] data);
    public byte[] SignData(ReadOnlySpan<byte> data);
    public int SignData(ReadOnlySpan<byte> data, Span<byte> destination);
    public bool TrySignData(ReadOnlySpan<byte> data, Span<byte> destination, out int bytesWritten);
    protected abstract int SignDataCore(ReadOnlySpan<byte> data, Span<byte> destination);

    public bool VerifyData(byte[] data, byte[] signature);
    public bool VerifyData(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature);
    protected abstract bool VerifyDataCore(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature);
}
public sealed partial class Ed25519OpenSsl : Ed25519
{
    [UnsupportedOSPlatformGuard("android")]
    [UnsupportedOSPlatformGuard("browser")]
    [UnsupportedOSPlatformGuard("ios")]
    [UnsupportedOSPlatformGuard("tvos")]
    [UnsupportedOSPlatformGuard("windows")]
    public static new bool IsSupported { get; }

    [UnsupportedOSPlatform("android")]
    [UnsupportedOSPlatform("browser")]
    [UnsupportedOSPlatform("ios")]
    [UnsupportedOSPlatform("tvos")]
    [UnsupportedOSPlatform("windows")]
    public Ed25519OpenSsl(SafeEvpPKeyHandle pkeyHandle);

    [UnsupportedOSPlatform("android")]
    [UnsupportedOSPlatform("browser")]
    [UnsupportedOSPlatform("ios")]
    [UnsupportedOSPlatform("tvos")]
    [UnsupportedOSPlatform("windows")]
    public Ed25519OpenSsl();
}
```

```C#
namespace System.Security.Cryptography.X509Certificates;

public sealed partial class PublicKey
{
    public Ed25519? GetEd25519PublicKey();
}

public partial class X509Certificate2
{
    public Ed25519? GetEd25519PublicKey();
    public Ed25519? GetEd25519PrivateKey();
    public X509Certificate2 CopyWithPrivateKey(Ed25519 privateKey);
}
```

**Comment by Fabi** [comment]
Do you keep the different Ed25519 modes/variants in mind?

**Comment by vcsjones** [comment]
> Do you keep the different Ed25519 modes/variants in mind?

The Ed25519ph and Ed25519ctx variants of Ed25519 are currently out of scope of this API proposal. We know they exist, but the Apple platform doesn't expose these variants so we could not reasonably implement them for anything except Unix with OpenSSL. That doesn't mean "no we won't ever do it", just that (Pure) Ed25519 is the current priority.

If we are ever able to implement them, then we have a reasonable idea of what the API shape will look like to accommodate them.

**Comment by annerajb** [comment]
@vcsjones

what's the next step now that it's approved?
Do I need to update the branch I had a year ago to match this spec / api?
Is there other dependent work needed for example for macos?
Where should the branch be based off main?

**Comment by vcsjones** [comment]
> what's the next step now that it's approved?

I have a lot of it implemented at https://github.com/vcsjones/runtime/tree/ed25519-impl.

So, largely the plan is for me to just finish that and get it merged soon and it will eventually flow in to a .NET 8 preview.

**Comment by vcsjones** [comment]
In https://github.com/dotnet/runtime/issues/20342#issuecomment-1540668616, we modified the approved API here to remove `SupportedOSPlatform{Guard}` attributes. The approved API has been modified to reflect that.

**Comment by vcsjones** [comment]
After working through the implementation and some of the finer details of tests against test vectors,
I've encountered a peculiar behavior of Apple's implementation of Ed25519 with CryptoKit. Specifically, it does not produce deterministic signatures.
Apple's implementation produces random signatures. These signatures made by Apple are Ed25519 "compatible", meaning that
signatures CryptoKit produces can be verified by OpenSSL (and other Ed25519 implementations), but results in failures against test vectors when checking the deterministic signature results. I looked around [Apple's documentation][1]
which lead me to this sentence buried in their documentation.

> Although not required by RFC 8032, which describes the Edwards-Curve Digital Signature Algorithm (EdDSA), the CryptoKit
> implementation of the algorithm employs randomization to generate a different signature on every call, even for the same
> data and key, to guard against side-channel attacks.

So it's not an incorrect use of their API, this is by design. Apple's wording here is a bit odd though, as they make it seem like an option. Specifically, RFC 8032 states:

> EdDSA signatures are deterministic.

There doesn't seem to be a lot of room for interpretation there. The RFC says they are deterministic. Apple's are not.
For good measure, I read what [FIPS 186-5][2] had to say:

> EdDSA signatures are deterministic. The signature is generated using the hash of the private key and the message
> using the procedure below or an equivalent process.

Again, little room for interpretation. Apple's implementation seemingly does what the RFC and the NIST publication say not
to do.

This more or less leads to the conclusion that Apple's Ed25519 is "Ed25519-like", but it is not the algorithm as described
in the RFC or the FIPS publication.

This puts us in a tough position. Our stance is we should be implementing Ed25519 on macOS. Not something that resembles
Ed25519.

I can think of a few ways forward here. For the sake of completeness, I will list everything that crossed my mind.

1. We ignore it and continue as-is. Apple chose to implement their implementation in a non-conforming way. We are just
    doing what the platform is telling us to do.

2. We split the difference: we can proceed with verify-only on Apple. Producing signatures would not function.

3. We drop Apple as a platform and proceed only with OpenSSL.

3. We delay Ed25519 in .NET until we have at least two spec-conforming implementations available for use on two different platforms.

I do not like one for reasons I already hinted at: even well intentioned, Apple's implementation is not "real" Ed25519. They
are generating a nonce in a non-deterministic way, which two specifications say not to do it that way. We would not, for example, be able to claim that our implementation follows an RFC or NIST design.

I don't like number two, because it's a half-measure and it too closely resembles number three.

For number three, this goes against the rule-of-two requirement for crypto in .NET. We need at least two implementations
to have confidence in the API shape, compatibility, and interoperability.

That leaves us with number four, which is unfortunately what I think is the right thing to do. If we had other platforms
lined up to support Ed25519, we could probably continue without Apple. But we don't. Neither Windows nor Android give us an Ed25519 primitive. Without Apple, that leaves us with just OpenSSL, and as previously stated, we don't like new crypto
primitives that support only one implementation.

@bartonjs I would be curious about your thoughts on the matter and if I am just thinking too hard about this.


[1]: https://developer.apple.com/documentation/cryptokit/curve25519/signing/privatekey/signature(for:)
[2]: https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf

**Comment by filipnavara** [comment]
I think the nonce randomization comes from the Ed25519 paper (https://ed25519.cr.yp.to/ed25519-20110926.pdf) and predates both of the EdDSA standardization efforts. (Then again, I am on vacation now, so I am too lazy to check the details; but I vaguely remember it so I thought I would post a reference.)

**Comment by vcsjones** [comment]
> I think the nonce randomization comes from the Ed25519 paper (https://ed25519.cr.yp.to/ed25519-20110926.pdf) and predates both of the EdDSA standardization efforts.

The paper explicitly says that the nonce (`r`) is not random:

> EdDSA avoids these issues by generating r = H(hb,...,h2b−1,M), so that different messages will lead to different, hard-to-predict values of r. No per- message randomness is consumed. It was standardized more or less as it was designed, including the deterministic signatures.

Apple is employing what's referred to these days as a synthetic nonce, where they use a hash over the secret key, the message, _and_ randomness. If the randomness fails entirely (the famous ECDSA k issue) then the scheme is still secure.

I understand why Apple has done this, and some notable members of the cryptographic community think synthetic nonces are better, but it's still an intentional deviation from the original design and the specs.

**Comment by bartonjs** [comment]
Huh.

I just went through and checked RFC errata to see if they've made a technical addendum that says non-deterministic signatures are OK.  There aren't.  In fact, both documents double-down on saying that Apple is wrong here:

>    EdDSA signatures are deterministic.  This protects against attacks
   arising from signing with bad randomness; the effects of which can,
   depending on the algorithm, range up to full private key compromise.
   It can be surprisingly hard to ensure good-quality random numbers,
   and there have been numerous security failures relating to this.
>
>  -- [IETF RFC 8032 Section 8.2](https://datatracker.ietf.org/doc/html/rfc8032#section-8.2)

> EdDSA signatures are deterministic; ... This process protects against attacks arising from generating signatures with insufficient randomness for the per-message secret number.
>
> -- [FIPS 186-5](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf) Section 7.1

And Apple is apparently defending against _defending against_ random numbers.

Another half-option occurred to me, which would come in at "1.5", I think: Adding a property like `public bool IsDeterministic { get; }`... but that could just be renamed to `public bool IsConformantImplementation { get; }`, and once we're saying "this is a thing that's like, but not actually, Ed25519" that just sounds like a mistake.

I had another flash of hope... I went to see if Apple got a CAVP/NIST certificate of conformance, and if it had any weasel words attached.  But, it turns out, CAVP doesn't yet test Ed25519 (or, at least, search doesn't list it as a searchable algorithm): https://csrc.nist.gov/projects/cryptographic-module-validation-program/validated-modules/search?SearchMode=Advanced&CertificateStatus=Active&ValidationYear=0&OtherAlgorithms=ed25519

So... yeah, I think I agree with (4).  We shouldn't proceed until at least one of:

* Apple adds a "be conformant" option
* Windows adds Ed25519 support (and it conforms)
* Android adds Ed25519 support (and it conforms)
* NIST throws out a 186-6 or an SP saying that non-deterministic variants of Ed25519 are now conformant.
  * Yes, NIST, not just IETF/IRTF.  We have FIPS pass-through concerns.
* CAVP starts doing Ed25519, issues Apple a certificate, and we all get to stare at it and learn "non-conformant is conformant after all", or what weasel words CAVP decides to put on the cert.

**Comment by filipnavara** [comment]
Thanks both for double-checking 👍

**Comment by Blackclaws** [comment]
> I do not like one for reasons I already hinted at: even well intentioned, Apple's implementation is not "real" Ed25519. They
> are generating a nonce in a non-deterministic way, which two specifications say not to do it that way. We would not, for example, be able to claim that our implementation follows an RFC or NIST design.

I think that in the end it comes down to what the end goal is. For me as a user of the implementation I only care about the signatures being compatible between implementations. If the signatures produced on Apple are verifiable by other EdDsa implementations and don't produce invalid signatures those would suffice in my opinion for anything I would use them for (jwt token generation etc.).

It might be that in some cases where government contracts or the likes are affected people need FIPS compliant implementations but iirc those are sometimes actually worse than state of the art implementations of common crypto.

Seeing as how Apple doesn't provide an alternative in this case and the only alternative is non-implementation I'd say go ahead but mark it explicitly as non-conformant for Apple platforms with a big warning sign.

**Comment by bartonjs** [comment]
> mark it explicitly as non-conformant for Apple platforms with a big warning sign.

How/where would you recommend we do that?  If someone wrote a platform neutral library that used Ed25519 so long as `Ed25519.IsSupported` returned `true`, that library would now need to carry the viral "if you run me on macOS I won't produce answers that necessarily pass all tests".

We'd essentially have to make a clone of the UnsupportedOS analyzer to carry something like NonConformantImplementation. Even that's not really good enough, because annotations are a design time problem, and conformance is a runtime problem... if .NET 8 says that Ed25519 is non-conformant on macOS, then any library building against .NET 8 as a minimum version believes that to be true.  But, by .NET 9, we learn that Apple has pushed out an update to make their algorithm conform.  Well, we can't remove the annotation from the .NET 8 targeting pack... so now the annotations lie.

While the RFC doesn't literally say `EdDSA signatures MUST be deterministic`, it's a MUST-level statement in it; so Apple's algorithm isn't actually EdDSA and thus isn't actually Ed25519, so we shouldn't call it Ed25519 in .NET.  We could call it `Ed25519SignatureCompatible`... but that's clearly just an attempt to ship _something_.

Another problem that I'd have with "go ahead" here is that we try really hard to make .NET be .NET everywhere.  When we can't, we usually want the experience to be PlatformNotSupportedException (or a similar "sorry, this doesn't work").  Saying that the same code gives a deterministic signature on some OSes, and a non-deterministic signature on others, isn't staying in that spirit.  So, yes, while the alternative is non-implementation, I feel that non-implementation is the better option.

**Comment by Blackclaws** [comment]
> How/where would you recommend we do that?

Mostly documentation regarding this function. In the end anyone that uses it should have read the documentation and know what they are getting themselves into. I get that this might be an issue for libraries that are cross platform and just link back to the base implementation and as you say the non-conformancy should spread to those sort of.



> Well, we can't remove the annotation from the .NET 8 targeting pack... so now the annotations lie.

Might it be a better solution to not push this directly into the runtime but instead have it as a NuGet package that stays in preview until this issue is resolved? That way people that know what they are getting themselves into can use it but it doesn't appear as readily available to the general public.


> Another problem that I'd have with "go ahead" here is that we try really hard to make .NET be .NET everywhere.

I think this is the biggest argument in favor of non-implementation. However there are enough things still in .NET that behave differently on Windows vs. other platforms and others that don't work the same on macOS either (see HTTP1/2 over HTTPS) which don't necessarily lead to PlatformNotSupported but just slightly different behaviour. 

In the end I get that you'd not want to add to the pile of things that work differently though so fair enough. To me it seems that unless you'd consider shipping just the Linux Openssl implementation this might stay in Limbo for a rather long time. Is there any specific reason why a macOS Openssl version wouldn't work? Is it just that this is a dependency you don't want to pull into the runtime?

**Comment by vcsjones** [comment]
> Is there any specific reason why a macOS Openssl version wouldn't work? Is it just that this is a dependency you don't want to pull into the runtime?

Two things here.

First, when I say "platform", I mean different cryptographic implementations, not operating systems. OpenSSL on macOS and Linux is still one "platform", so that does not help us with our desire to support multiple platforms with cryptographic primitives.

Second, shipping OpenSSL in-box is not an option - .NET's current stance, by policy, is that we do not ship implementations with cryptographic implementations. We have in the past used OpenSSL on macOS if happened to be on the system - this is still used for AesCcm, for example. However this is an anti-goal. We want to cut ties from OpenSSL on macOS, not take further dependencies on it.

**Comment by ranma42** [comment]
Would it make sense to only support signature verification? (maybe not, AFAICT the current interfaces expect that either both signing and verification are supported, or none of them is)

**Comment by jeffhandley** [comment]
Is there value in verify-only across the board? What are some scenarios for verifying but not producing?

**Comment by vcsjones** [comment]
> What are some scenarios for verifying but not producing?

The most common use case for verifying is when you want to verify a signature another party produced. JWT is a good example - a server may respond with a signed JWT, and a client has the public key for that party and wants to verify the signature exists.

I don't doubt there are many use cases for a verify only implementation - but in my personal opinion, it still teeters too close to having only one implementation.

**Comment by krwq** [comment]
Another option here could be to by default throw PNSE for non conformant stuff and create a feature switch "AllowNonConformantEd25519" - PNSE could provide detail about existence of that switch. I'm personally also fine with not supporting this until Windows/Android add their implementations

**Comment by vcsjones** [comment]
> Another option here could be to by default throw PNSE for non conformant stuff and create a feature switch "AllowNonConformantEd25519"

That more or less I think falls in line with something @bartonjs reasoned through:

> Adding a property like public bool IsDeterministic { get; }... but that could just be renamed to public bool IsConformantImplementation { get; }, and once we're saying "this is a thing that's like, but not actually, Ed25519" that just sounds like a mistake.

**Comment by Neustradamus** [comment]
To follow

**Comment by dariogriffo** [comment]
Any update on this? At the moment we rely on a nuget package based on bouncy castle to sign our jwr bit is outdated and still references microsoft.jwt.model 5.4.0 with critical unresolved security issues

**Comment by vcsjones** [comment]
> Any update on this?

Unfortunately not. The status remains the same as noted in [this previous comment](https://github.com/dotnet/runtime/issues/63174#issuecomment-1563316891). The implementation remains blocked.

**Comment by Fabi** [comment]
Then again go damn back to openssl used crypto stuff. You guys block so many things on mac or functionality in general just because you don't want to use openssl on mac for example. It worked before (see aes gcm) where it was removed for now reasons too and replaced with the apple implementation for no real reasons.

**Comment by filipnavara** [comment]
OpenSSL is not distributed in-box with operating system on macOS. That would shift the burden of distributing it to you as a user, including all security updates.

**Comment by filipnavara** [comment]
> Did you really just suggest using another 3rd party lib that has a dep on another 3rd party library...

No. NSec is port of libsodium to C#, no extra dependencies and it is self-contained. It's merely a suggestion, not an endorsement. As much as I would like .NET to ship an in-box implementation of this cryptographic primitive it's not currently possible within the given constrains (cryptography stack is shipped with OS and .NET doesn't implement encryption algorithms by itself; at least two of the three major platforms support a given cryptographic primitive).

> ...instead of having people just use a simple brew install openssl or so on mac which is a very common thing, especially for devs.

Firstly, you somehow assume that all consumers of .NET are developers. That's not the case. Neither is `brew` installed on macOS machines by default, and it doesn't come with a built-in update mechanism for security updates either. That makes it a non-starter as a general purpose solution.

> It worked fine for everyone before when some things used openssl on mac

It didn't work fine for everyone. In fact, it was a common pit of failure because developers often had OpenSSL installed on their machines and mistakenly assumed that `AesGcm` universally works, and they came in for a surprise once they started testing their application on customer machines.

**Comment by Fabi** [comment]
NSec pinvokes libsodium so it is a dependency and not a port.

I don't think there were many open issues anywhere (can't speak for you doing internal stuff) regarding openssl and issues on it.

Sure brew is not installed by default but it's still the most common way to install any kind of packages like openssl or others. Including updating those. It's very rare that you use other things on macOS regarding that.

Also why not keep it as an option as proposed before through an environment variable to use openssl? That would have been nice for example for things like AesGcm which is fully unusable for me now on macOS as .NET "implementation" due to how a client that I have to work with. Sure I got a 'workaround' by copying the code before it got changed but still that wasn't nice to do. Such a switch (more like the openssl lib path) kind of worked at that time and turned AesGcm from unsupported to supported already.

**Comment by filipnavara** [comment]
> NSec pinvokes libsodium so it is a dependency and not a port.

I stand corrected. There was some other managed port and I got them confused. That said, it solves the distribution problem by shifting it into NuGets that are linked into your application and where you are responsible for their timely updates. libsodium was also audited several times in the past so it comes with some indirect guarantees. It may not be a perfect solution but it's a solution that you can use right now.

> Sure brew is not installed by default but it's still the most common way to install any kind of packages like openssl or others. 

I am not in any way disputing that, when it comes to **developers**. It's totally different story when it comes to end users who obtain most applications from the App Store or as web downloads. I happen to be a developer with `brew` and `openssl` on my machine, but I also distribute an end user application to millions of users among multiple platforms and that's vastly different from a scenario where you have total control of your machine or the machines where the software is deployed.

> Also why not keep it as an option as proposed before through an environment variable to use openssl? That would have been nice for example for things like AesGcm which is fully unusable for me now on macOS as .NET "implementation" due to how a client that I have to work with.

This is off-topic and we have already discussed this to death in the past. The decision to drop the OpenSSL support from `AesGcm` on macOS was done with the same pit-of-failure argument. I was not the one presenting the argument and I was not the one making the decision, I simply agree with it. You are free to disagree, open an issue, present your use case and ask for the decision to be re-evaluated.

**Comment by amirvenus** [comment]
Would be great if the supported includes .net Android as well.

**Comment by vcsjones** [comment]
Android 13 (API level 33) is documented as supporting Ed25519. https://developer.android.com/reference/kotlin/java/security/Signature

Let's see if we can land Ed25519 for Linux and Android.

**Comment by teo-tsirpanis** [comment]
For visibility, Microsoft recently said this about adding Ed25519 to Windows (https://github.com/microsoft/SymCrypt/issues/43#issuecomment-2578933331):

> historically Microsoft has disallowed the use of EdDSA in our products because it was not a FIPS-approved algorithm. As of 2023, EdDSA is now FIPS approved, but we're now in the process of implementing and adopting PQC algorithms such as ML-DSA, so it remains a low priority for us. As such, I do not expect that we'll implement EdDSA any time soon, if ever.

**Comment by ChadNedzlek** [comment]
I was also trying to use a protocol that requires this algorithm. So now my only choice is to ship some other random native library with my app.  I care zero percent about PQC. I care very much about "algorithms in wide usage right now".

**Comment by teo-tsirpanis** [comment]
@vcsjones should the class' API shape be reconsidered in line with the PQC algorithm classes or the proposed `X25519DiffieHellman` class?

**Comment by vcsjones** [comment]
@bartonjs I have updated the top issue with our new proposal.

**Comment by bartonjs** [comment]
Me as crypto owner: The updated proposal looks good, marked as ready for review.
Me as API Review chair: As this issue was long-dormant, it should be treated as a new proposal effective 2026-04-20T17:57-0700; which means it should sort just after BitArray/ReadOnlySpan, despite apireview.net sorting it at the top (it uses the original ready-for-review so that something being bounced back for more research comes back at the front, rather than at the back, so we can process the update while still having context)

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=Z3vqvBGpFrk&t=0h38m16s)

* Corrected typo in Ed25519OpenSsl..ctor
* Otherwise, looks good as proposed.

```csharp

namespace System.Security.Cryptography
{
    public abstract partial class Ed25519 : System.IDisposable
    {
        //  RFC 8032
        public const int SignatureSizeInBytes = 64;
        public const int PrivateKeySizeInBytes = 32;
        public const int PublicKeySizeInBytes = 32;

        protected Ed25519();

        public static bool IsSupported { get; }

        public static Ed25519 GenerateKey();

        public static Ed25519 ImportPublicKey(byte[] source);
        public static Ed25519 ImportPublicKey(ReadOnlySpan<byte> source);

        public static Ed25519 ImportPrivateKey(byte[] source);
        public static Ed25519 ImportPrivateKey(ReadOnlySpan<byte> source);

        public byte[] ExportPublicKey();
        public void ExportPublicKey(Span<byte> destination);
        protected abstract void ExportPublicKeyCore(Span<byte> destination);

        public byte[] ExportPrivateKey();
        public void ExportPrivateKey(Span<byte> destination);
        protected abstract void ExportPrivateKeyCore(Span<byte> destination);

        public byte[] SignData(byte[] data);
        public void SignData(ReadOnlySpan<byte> data, Span<byte> destination);
        protected abstract void SignDataCore(ReadOnlySpan<byte> data, Span<byte> destination);

        public bool VerifyData(byte[] data, byte[] signature);
        public bool VerifyData(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature);
        protected abstract bool VerifyDataCore(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature);

        public void Dispose();
        protected virtual void Dispose(bool disposing);

        public bool TryExportEncryptedPkcs8PrivateKey(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters, Span<byte> destination, out int bytesWritten);
        public bool TryExportEncryptedPkcs8PrivateKey(ReadOnlySpan<char> password, PbeParameters pbeParameters, Span<byte> destination, out int bytesWritten);
        public bool TryExportEncryptedPkcs8PrivateKey(string password, PbeParameters pbeParameters, Span<byte> destination, out int bytesWritten);

        public byte[] ExportEncryptedPkcs8PrivateKey(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters);
        public byte[] ExportEncryptedPkcs8PrivateKey(ReadOnlySpan<char> password, PbeParameters pbeParameters);
        public byte[] ExportEncryptedPkcs8PrivateKey(string password, PbeParameters pbeParameters);

        public string ExportEncryptedPkcs8PrivateKeyPem(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters);
        public string ExportEncryptedPkcs8PrivateKeyPem(ReadOnlySpan<char> password, PbeParameters pbeParameters);
        public string ExportEncryptedPkcs8PrivateKeyPem(string password, PbeParameters pbeParameters);

        public byte[] ExportPkcs8PrivateKey();
        public string ExportPkcs8PrivateKeyPem();
        public bool TryExportPkcs8PrivateKey(Span<byte> destination, out int bytesWritten);
        protected abstract bool TryExportPkcs8PrivateKeyCore(Span<byte> destination, out int bytesWritten);

        public byte[] ExportSubjectPublicKeyInfo();
        public bool TryExportSubjectPublicKeyInfo(Span<byte> destination, out int bytesWritten);
        public string ExportSubjectPublicKeyInfoPem();

        public static Ed25519 ImportEncryptedPkcs8PrivateKey(ReadOnlySpan<byte> passwordBytes, ReadOnlySpan<byte> source);
        public static Ed25519 ImportEncryptedPkcs8PrivateKey(ReadOnlySpan<char> password, ReadOnlySpan<byte> source);
        public static Ed25519 ImportEncryptedPkcs8PrivateKey(string password, byte[] source);

        public static Ed25519 ImportFromEncryptedPem(ReadOnlySpan<char> source, ReadOnlySpan<byte> passwordBytes);
        public static Ed25519 ImportFromEncryptedPem(string source, byte[] passwordBytes);
        public static Ed25519 ImportFromEncryptedPem(ReadOnlySpan<char> source, ReadOnlySpan<char> password);
        public static Ed25519 ImportFromEncryptedPem(string source, string password);

        public static Ed25519 ImportFromPem(ReadOnlySpan<char> source);
        public static Ed25519 ImportFromPem(string source);

        public static Ed25519 ImportPkcs8PrivateKey(byte[] source);
        public static Ed25519 ImportPkcs8PrivateKey(ReadOnlySpan<byte> source);

        public static Ed25519 ImportSubjectPublicKeyInfo(byte[] source);
        public static Ed25519 ImportSubjectPublicKeyInfo(ReadOnlySpan<byte> source);
    }

    public sealed partial class Ed25519OpenSsl : Ed25519
    {
        [UnsupportedOSPlatformAttribute("android")]
        [UnsupportedOSPlatformAttribute("browser")]
        [UnsupportedOSPlatformAttribute("ios")]
        [UnsupportedOSPlatformAttribute("osx")]
        [UnsupportedOSPlatformAttribute("tvos")]
        [UnsupportedOSPlatformAttribute("windows")]
        public Ed25519OpenSsl(SafeEvpPKeyHandle pkeyHandle);

        public SafeEvpPKeyHandle DuplicateKeyHandle();
    }
}
```
